ruma-api-macros/src/util.rs

```diff
@@ -0,0 +1,259 @@
+        quote! {
+            let request_query = match ruma_api::exports::ruma_serde::urlencoded::from_str(
+                &request.uri().query().unwrap_or("")
+            ) {
+                Ok(query) => query,
+                Err(err) => {
+                    return Err(
+                        ruma_api::error::RequestDeserializationError::new(err, request).into()
+                    );
+                }
+            };
+        }
+    } else if request.has_query_fields() {
+        quote! {
+            let request_query: RequestQuery =
+                match ruma_api::exports::ruma_serde::urlencoded::from_str(
+                    &request.uri().query().unwrap_or("")
+                ) {
+                    Ok(query) => query,
+                    Err(err) => {
+                        return Err(
+                            ruma_api::error::RequestDeserializationError::new(err, request)
+                                .into()
+                        );
+                    }
+                };
+        }
+    } else {
+        TokenStream::new()
+    }
+}
+
+/// Generates the code to initialize a `Request`.
+///
+/// Used to construct an `http::Request`s body.
+pub(crate) fn build_request_body(request: &Request) -> TokenStream {
+    if let Some(field) = request.newtype_raw_body_field() {
+        let field_name = field.ident.as_ref().expect("expected field to have an identifier");
+        quote!(request.#field_name)
+    } else if request.has_body_fields() || request.newtype_body_field().is_some() {
+        let request_body_initializers = if let Some(field) = request.newtype_body_field() {
+            let field_name = field.ident.as_ref().expect("expected field to have an identifier");
+            quote! { (request.#field_name) }
+        } else {
+            let initializers = request.request_body_init_fields();
+            quote! { { #initializers } }
+        };
+
+        quote! {
+            {
+                let request_body = RequestBody #request_body_initializers;
+                ruma_api::exports::serde_json::to_vec(&request_body)?
+            }
+        }
+    } else {
+        quote!(Vec::new())
+    }
+}
+
+pub(crate) fn parse_request_body(request: &Request) -> TokenStream {
+    if let Some(field) = request.newtype_body_field() {
+        let field_name = field.ident.as_ref().expect("expected field to have an identifier");
+        quote! {
+            #field_name: request_body.0,
+        }
+    } else if let Some(field) = request.newtype_raw_body_field() {
+        let field_name = field.ident.as_ref().expect("expected field to have an identifier");
+        quote! {
+            #field_name: request.into_body(),
+        }
+    } else {
+        request.request_init_body_fields()
+    }
+}
+
+pub(crate) fn req_res_meta_word<T>(
+    attr_kind: &str,
+    field: &syn::Field,
+    newtype_body_field: &mut Option<syn::Field>,
+    body_field_kind: fn() -> T,
+    raw_field_kind: fn() -> T,
```

Comment:
Why not just

```suggestion
    body_field_kind: T,
    raw_field_kind: T,
```

?

Reply:
:facepalm: Why did I make that complicated, Ooops.